pkg/storage/test/tuples.go

```diff
@@ -1179,511 +1138,172 @@ func ReadStartingWithUserTest(t *testing.T, datastore storage.OpenFGADatastore)
-		require.NotEqual(t, firstRead[0].GetKey().GetRelation(), secondRead[0].GetKey().GetRelation())
-		require.Empty(t, contToken)
-	})
-}
-
-func ReadPageTestCorrectnessOfTuples(t *testing.T, datastore storage.OpenFGADatastore) {
-	ctx := context.Background()
-
-	tuples := []*openfgav1.TupleKey{
-		tuple.NewTupleKey("document:1", "reader", "user:anne"),
-		tuple.NewTupleKey("document:1", "reader", "user:bob"),
-		tuple.NewTupleKey("document:1", "writer", "user:bob"),
-		{
-			Object:   "document:2",
-			Relation: "viewer",
-			User:     "user:anne",
-			Condition: &openfgav1.RelationshipCondition{
-				Name: "condition",
+	testCases := map[string]struct {
+		filter         *openfgav1.TupleKey
+		expectedTuples []*openfgav1.TupleKey
+	}{
+		`no_filter`: {
+			filter:         tuple.NewTupleKey("", "", ""),
+			expectedTuples: tuples,
+		},
+		`filter_by_user_and_relation_and_objectID`: {
+			filter: tuple.NewTupleKey("document:1", "reader", "user:github.com|anne@test.com"),
+			expectedTuples: []*openfgav1.TupleKey{
+				tuple.NewTupleKey("document:1", "reader", "user:github.com|anne@test.com"),
+			},
+		},
+		`filter_by_user_and_relation_and_objectType`: {
+			filter: tuple.NewTupleKey("document:", "reader", "user:github.com|bob@test.com"),
+			expectedTuples: []*openfgav1.TupleKey{
+				tuple.NewTupleKey("document:1", "reader", "user:github.com|bob@test.com"),
+			},
+		},
+		`filter_by_user_and_objectType`: {
+			filter: tuple.NewTupleKey("document:", "", "user:github.com|bob@test.com"),
+			expectedTuples: []*openfgav1.TupleKey{
+				tuple.NewTupleKey("document:1", "reader", "user:github.com|bob@test.com"),
+				tuple.NewTupleKey("document:1", "writer", "user:github.com|bob@test.com"),
+			},
+		},
+		`filter_by_relation_and_objectID`: {
+			filter: tuple.NewTupleKey("document:1", "reader", ""),
+			expectedTuples: []*openfgav1.TupleKey{
+				tuple.NewTupleKey("document:1", "reader", "user:github.com|anne@test.com"),
+				tuple.NewTupleKey("document:1", "reader", "user:github.com|bob@test.com"),
+			},
+		},
+		`filter_by_objectID`: {
+			filter: tuple.NewTupleKey("document:1", "", ""),
+			expectedTuples: []*openfgav1.TupleKey{
+				tuple.NewTupleKey("document:1", "reader", "user:github.com|anne@test.com"),
+				tuple.NewTupleKey("document:1", "reader", "user:github.com|bob@test.com"),
+				tuple.NewTupleKey("document:1", "writer", "user:github.com|bob@test.com"),
+				tuple.NewTupleKey("document:1", "admin", "user:github.com|anne@test.com"),
+				tuple.NewTupleKeyWithCondition("document:1", "viewer", "user:github.com|anne@test.com", "condition1", nil),
+			},
+		},
+		`filter_by_objectID_and_user`: {
+			filter: tuple.NewTupleKey("document:1", "", "user:github.com|bob@test.com"),
+			expectedTuples: []*openfgav1.TupleKey{
+				tuple.NewTupleKey("document:1", "reader", "user:github.com|bob@test.com"),
+				tuple.NewTupleKey("document:1", "writer", "user:github.com|bob@test.com"),
+			},
+		},
+		`filter_by_objectID_with_special_character`: {
```

Comment:
Did we drop filter_by_object_type_and_user_with_special_character?

Reply:
No, it's the test `filter_by_objectID_with_special_character`. Notice how both user and object have `|` 😉

Reply:
I mean the filter part.  Here, we only filter by document that has special character.  Do we have any request that filter by type + user that has special character?

Reply:
never mind I see the test above.